### PR TITLE
bpo-47115: Document which parts of structs are in limited API/stable ABI

### PR DESCRIPTION
**Pull request by encukou**
Some screenshots of the result:

Opaque struct:
image
Fully visible structs:
image
image
Partial struct (see the issue for details):
image
Included member of the above:
image

Non-struct items don't change:
image

https://bugs.python.org/issue47115

### CODE REVIEW

File: Misc/stable_abi.txt
Comment (vstinner) [inline Misc/stable_abi.txt:105]: By the way, I recently added "The structure is not part of the C API." to:
https://docs.python.org/dev/c-api/frame.html

Comment (encukou) [reply Misc/stable_abi.txt:105]: That's wrong. If it's not part of the API at all, you can't use functions like `PyFrame_GetBack` that take it as an argument.

**Comment by encukou** [comment]
> Is there a definition of an "opaque structure" in the Python doc? Is there a link from "opaque struct" in then generated doc?

IMO it's a generally known term, it even has a [Wikipedia entry](https://en.wikipedia.org/wiki/Opaque_data_type). I don't think the docs need to define it.

**Comment by encukou** [comment]
@arhadthedev, I'll open an issue for it.

File: Misc/stable_abi.txt
Comment (erlend-aasland) [inline Misc/stable_abi.txt]:
```suggestion
#   the Stable ABI, and must not change.
```

File: Misc/stable_abi.txt
Comment (erlend-aasland) [inline Misc/stable_abi.txt]: I found that a little hard to read. Maybe something like this?

```suggestion
# - fields: Space-separated list of fields which are part of the Limited API
```

Nit: I also wonder if it would be more consistent to name it `members` iso. `fields`.

**Review by erlend-aasland — COMMENTED** [review]
Nice; I think this doc enhancement will be helpful for both core devs and others.

From what I could see, the Py changes look sensible, but I'm not at all familiar with these doc tools. I'll have another look at this tomorrow.

File: Misc/stable_abi.txt
Comment (vstinner) [reply Misc/stable_abi.txt:105]: It seems like there is a misunderstanding on what "a structure" means here. Python C API is misleading because it uses short ``PyFrameObject`` type name rather than ``struct PyFrameObject``.

In practice in C, "PyFrameObject" is a type name, not a structure. I recently added ``Include/pytypedefs.h`` to clarify which type names are always available in the C API (and fix a compiler error), even if their related structures are opaque. I understand that the **type** (type "name") is part of the C API, but the structure is not.

I'm not sure how  https://docs.python.org/dev/c-api/frame.html should be rephrased. The follow this PR, maybe "The structure is opaque" which may become redundant once this PR is merged.

Comment (encukou) [reply Misc/stable_abi.txt:105]: Ah, I see.
This should be explained/avoided:  if you don't know the term "opaque structure" you are likely to search for it, but if you don't realize the distinction between "type" and "structure", you'll probably just be confused.

I'll work on the rephrasing next week. Please let me know if you know of other places with similar wording.

Comment (encukou) [reply Misc/stable_abi.txt:105]: I started going through all the structs to make sure the docs make sense.
And I quite like the note for [PyInterpreterState](https://docs.python.org/3/c-api/init.html#c.PyInterpreterState): “There are no public members in this structure.” Short and clear! I'll use that for PyFrameObject too.

**Comment by encukou** [comment]
So, how about this?
image
(The “structure of the objects used to describe frame objects” is weird, but I'm leaving that out of scope of this PR.)

**Comment by vstinner** [comment]
“There are no public members in this structure.”

That's explicit. It prevents the confusion about "the structure is not part of the C API" and confusion between "structure" and "type". In practice, an opaque structure ... is an empty structure, right ;-) The only subtle difference is that sizeof() doesn't work on an opaque structure, but I don't think that it's worth it to do into the details, unless we add an entry to the glossary for "opaque structure".

**Review by vstinner — APPROVED** [review]
LGTM.

**Review by erlend-aasland — APPROVED** [review]
Looks good to me also.

It looks to me that this comment is now outdated:

https://github.com/python/cpython/blob/4b188ae910dc65b5d6df3b2c5f0f826a91900dd9/Doc/tools/extensions/c_annotations.py#L101-L103

Either update it or delete it :)

**Comment by encukou** [comment]
Comment adjusted, and one more tweak to the wording:
image

(It should be possible to make these -- PyObject & PyVarObject -- fully opaque in the 3.12 limited API.)

**Pull request merged**
